clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -845,6 +845,12 @@ void CodeGenFunction::StartFunction(GlobalDecl GD, QualType RetTy,
   if (SanOpts.has(SanitizerKind::ShadowCallStack))
     Fn->addFnAttr(llvm::Attribute::ShadowCallStack);
 
+  if (SanOpts.has(SanitizerKind::Realtime)) {
+    for (const FunctionEffectWithCondition &Fe : FD->getFunctionEffects())
+      if (Fe.Effect.kind() == FunctionEffect::Kind::NonBlocking)
+        Fn->addFnAttr(llvm::Attribute::SanitizeRealtime);
+  }
```

Comment:
@dougsonos , would love some eyes on this to make sure this makes sense to you!

Reply:
`getFunctionEffects()` is cheap, but when investigating the performance regression introduced by effects, it turned out to be worthwhile to test `Context.hasAnyFunctionEffects()` first, to optimize for the case of no effects ever having been seen.

Reply:
Excellent, added that check in there.